1. Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks):
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].
Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с помощью функции cov из numpy
Полученные значения должны быть равны.
Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков,
а затем с использованием функций из библиотек numpy и pandas.


In [25]:
import pandas as pd
import numpy as np
from scipy import stats

In [2]:
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]

In [9]:
MX = sum(zp)/len(zp)
MY = sum(ks)/len(ks)

In [12]:
s = 0
for i in range(len(zp)):
    s += (zp[i]- MX)*(ks[i]-MY)
s
    

91578.4

In [13]:
cov = s / len(zp)
cov

9157.84

In [15]:
np.cov(zp,ks,ddof=0)

array([[ 3494.64,  9157.84],
       [ 9157.84, 30468.89]])

Несмещенная ковариация

In [19]:
cov_n = s/(len(zp)-1)
cov_n

10175.377777777778

In [20]:
np.cov(zp,ks,ddof=1)

array([[ 3882.93333333, 10175.37777778],
       [10175.37777778, 33854.32222222]])

коэффициент корреляции Пирсона

In [21]:
DX = sum((x - MX)**2 for x in zp) / (len(zp) - 1)
DY = sum((y - MY)**2 for y in ks) / (len(zp) - 1)

In [23]:
sigm_X = DX ** 0.5
sigm_Y = DY ** 0.5

r = cov_n / (sigm_X * sigm_Y)
r

0.8874900920739162

In [24]:
np.corrcoef(zp, ks, ddof=1)

/var/folders/jg/q12qqgm14jx4nld88b6cdy3m0000gn/T/ipykernel_11774/3485982831.py:1: DeprecationWarning: bias and ddof have no effect and are deprecated
  np.corrcoef(zp, ks, ddof=1)


array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

In [26]:
stats.pearsonr(zp,ks)

(0.8874900920739164, 0.0006107546587257491)

2. Измерены значения IQ выборки студентов,
обучающихся в местных технических вузах:
131, 125, 115, 122, 131, 115, 107, 99, 125, 111.
Известно, что в генеральной совокупности IQ распределен нормально.
Найдите доверительный интервал для математического ожидания с надежностью 0.95.


In [29]:
iq = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])

In [30]:
p = 0.95

n = len(iq)
mean = iq.mean()
sigm = iq.std(ddof=1)
alpha = 1 - p

n, mean, sigm

(10, 118.1, 10.54566788359614)

In [31]:
t = stats.t.ppf(1 - alpha/2, df=n-1)
t

2.2621571627409915

In [32]:
L = mean - t * (sigm / np.sqrt(n))
U = mean + t * (sigm / np.sqrt(n))

L, U

(110.55608365158724, 125.64391634841274)

доверительный интервал для математического ожидания с надежностью 0.95 (110.55608365158724 <= a <= 125.64391634841274)

3. Известно, что рост футболистов в сборной распределен нормально
с дисперсией генеральной совокупности, равной 25 кв.см. Объем выборки равен 27,
среднее выборочное составляет 174.2. Найдите доверительный интервал для математического
ожидания с надежностью 0.95.


In [33]:
n = 27
p = 0.95
D = 25
mean = 174.2

sigm = D ** 0.5
alpha = 1 - p

t = stats.norm.ppf(1 - alpha/2)
t

1.959963984540054

In [34]:
L = mean - t * (sigm / np.sqrt(n))
U = mean + t * (sigm / np.sqrt(n))

L, U

(172.3140237765397, 176.08597622346028)

доверительный интервал для математического ожидания с надежностью 0.95 (172.3140237765397 <= a <= 176.08597622346028)